src/k_smallest.rs

```diff
@@ -1,20 +1,92 @@
-use alloc::collections::BinaryHeap;
-use core::cmp::Ord;
+use alloc::vec::{Vec, IntoIter};
+use core::cmp::{Ordering, Reverse};
 
-pub(crate) fn k_smallest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
-    if k == 0 { return BinaryHeap::new(); }
+/// Consumes a given iterator, returning the minimum elements in **ascending** order.
+pub(crate) fn k_smallest_general<T, I: Iterator<Item = T>>(
+    mut iter: I,
+    k: usize,
+    mut comparator: impl FnMut(&T, &T) -> Ordering,
+) -> IntoIter<T> {
+    if k == 0 {
+        return Vec::new().into_iter();
+    }
+    let mut storage: Vec<T> = iter.by_ref().take(k).collect();
+    let mut heap = &mut storage[..];
```

Comment:
If I'm reading this correctly, the largest element goes to `heap[0]`. We should comment this somewhere.

Reply:
> Could we scope `storage` and `heap` as narrow as possible? `storage` needed throughout the function, `heap` only for the sorting part below => scope `heap` as narrow as possible. (And possibly rename `storage` -> `heap`, `heap` -> `heap_suffix` or similar.)

I see what you mean about the scope, that should be an easy change. I'm less sure about the naming, because I think my original reasoning was that by the point `heap` shrinks to nothing, `storage` is *not* a heap, it's in strict order. (It also shrinks to the left, so it'd be `heap_prefix`) 

> If I'm reading this correctly, the largest element goes to `heap[0]`. We should comment this somewhere.

I had to check this carefully and I'm still not sure, so I'm definitely going to comment that somewhere or ideally, name the comparator something more suggestive.

Reply:
Scope has been shrunk, names changed, and lots more comments added so there's more explanation of what's going on. Hope that helps